### PR TITLE
OCLOMRS-393 : Add mapping for existing concept

### PR DESCRIPTION
**Pull request by Akayeshmantha**
# JIRA TICKET NAME:
[implement add mappings to existing concepts](https://issues.openmrs.org/browse/OCLOMRS-393)

# Summary:

Currently, mappings cannot be added to existing concepts, this has to be implemented.This is because some missing props for createform.jsx component.

### CODE REVIEW

**Comment by Akayeshmantha** [comment]
@Mrshinezee @collinewait @samdiano @abulojoshua1 @brucemakallan @AGMETEOR opened this new repo with the new branching strategy guys sorry for the previous mistake

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 1]
```diff
@@ -154,6 +162,47 @@ describe('Test suite for dictionary concepts components', () => {
     expect(notify.show).toHaveBeenCalledWith('enter a valid uuid', 'error', 3000);
   });
 
+  it('should call addMappingRow function', () => {
+    const wrapper = mount(<Router>
+      <EditConcept {...props} />
+    </Router>);
+    const instance = wrapper.find('EditConcept').instance();
+    instance.addMappingRow();
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: @Akayeshmantha yes. Without an assertion/expect then it's not a test. It's just an it() function running addMappingRow() function from the component instance. We can probably expect that something in the dom was added or there was a change in the state.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 2]
```diff
@@ -154,6 +162,47 @@ describe('Test suite for dictionary concepts components', () => {
...
+  });
+
+  it('should call removeMappingRow function', () => {
+    const wrapper = mount(<Router>
+      <EditConcept {...props} />
+    </Router>);
+    const event = { target: 0 };
+    const instance = wrapper.find('EditConcept').instance();
+    instance.removeMappingRow(event);
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: @Akayeshmantha yes.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 3]
```diff
@@ -154,6 +162,47 @@ describe('Test suite for dictionary concepts components', () => {
...
+  });
+
+  it('should call updateAsyncSelectValue function', () => {
+    const wrapper = mount(<Router>
+      <EditConcept {...props} />
+    </Router>);
+    const value = { index: 0, value: 'malaria 1', label: 'malaria 1' };
+    const instance = wrapper.find('EditConcept').instance();
+    instance.updateAsyncSelectValue(value);
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: @Akayeshmantha yes.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 4]
```diff
@@ -154,6 +162,47 @@ describe('Test suite for dictionary concepts components', () => {
...
+    </Router>);
+    const event = {
+      target: {
+        tabIndex: 0,
+        name: 'to_concept_name',
+        value: 'malaria',
+      },
+    };
+    const instance = wrapper.find('EditConcept').instance();
+    instance.updateEventListener(event);
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: @Akayeshmantha yes

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 5]
```diff
@@ -173,6 +222,17 @@ describe('Test suite for dictionary concepts components', () => {
...
+  it('it should render with a null conceptType', () => {
+    const newProps = {
+      ...props,
+    };
+    newProps.match.params.conceptType = null;
+    const NewWrapper = mount(<Router>
+      <EditConcept {...newProps} />
+    </Router>);
+    const instance = NewWrapper.find('EditConcept').instance();
+    instance.addMappingRow();
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: @Akayeshmantha yes. Without an assertion/expect then it's not a test. It's just an it() function running addMappingRow() function from the component instance. We can probably expect that something in the dom was added or there was a change in the state after that function has been called.

File: src/components/dictionaryConcepts/containers/EditConcept.jsx [hunk 6]
```diff
@@ -104,9 +115,9 @@ export class EditConcept extends Component {
     event.preventDefault();
   }
 
-  removeNewName(event, uuid) {
```
Comment (dkayiwa) [inline src/components/dictionaryConcepts/containers/EditConcept.jsx]: Why the change from uuid to id?

File: src/tests/Title.test.js [hunk 7]
```diff
@@ -3,8 +3,12 @@ import { mount } from 'enzyme';
 import Title from '../components/Title';
 
 describe('Title component', () => {
+  const props = {
```
Comment (dkayiwa) [inline src/tests/Title.test.js]: Is this related to the ticket?

Comment (Akayeshmantha) [reply src/tests/Title.test.js]: @dkayiwa  in order to get the test coverage up and pass from coverall had to write some unit tests apart from the ticket after talking wih @abulojoshua1.

File: src/components/dictionaryConcepts/containers/EditConcept.jsx (hunk 6 above)
Comment (Akayeshmantha) [reply src/components/dictionaryConcepts/containers/EditConcept.jsx]: since its uuid is already defined in the scope

File: src/tests/Title.test.js (hunk 7 above)
Comment (dkayiwa) [reply src/tests/Title.test.js]: Was the coverage drop caused by changes in this pull request?

Comment (Akayeshmantha) [reply src/tests/Title.test.js]: Nope u can check that in coveralls

Comment (dkayiwa) [reply src/tests/Title.test.js]: This pull request is about only one thing. And that is: Add mapping for existing concept.
Any other issues should have separate tickets and pull requests.

Comment (Akayeshmantha) [reply src/tests/Title.test.js]: @dkayiwa actuallly htis is only covering  Add mapping for existing concept.But becuase of the coverage drop @abulojoshua1  asked to add more unit tests

Comment (dkayiwa) [reply src/tests/Title.test.js]: If the tests are not covering what this ticket is all about, then they are in a wrong pull request.

Comment (Akayeshmantha) [reply src/tests/Title.test.js]: @dkayiwa tests covers all in the story.And as well as it tries to address coverage drop with new unit tests

Comment (dkayiwa) [reply src/tests/Title.test.js]: Just to give you one example, how is this test related to? "Add mapping for existing concept "

Comment (Akayeshmantha) [reply src/tests/Title.test.js]: @dkayiwa  done changes Daniel

File: src/components/dictionaryConcepts/containers/EditConcept.jsx (hunk 6 above)
Comment (Akayeshmantha) [reply src/components/dictionaryConcepts/containers/EditConcept.jsx]: import uuid from 'uuid/v4';
importing uuid module and it's already define so for a local parameter to a function thought of changing it to id

**Comment by dkayiwa** [comment]
Thank you so much @Akayeshmantha for the great work! 👍 
We are please so sorry for having had a duplicate ticket which ended up with this: https://github.com/openmrs/openmrs-ocl-client/pull/323

Nevertheless, we shall always remember your awesomeness towards contributing to this ticket! 😊
